File: examples/02-reliability/python/async_procedure_and_race.py
import asyncio
from datetime import timedelta

from temporalio import workflow

# `send`, `ask_primary`, `ask_fallback` are module-level @activity.defn functions.


@workflow.defn
class AsyncProcedureAndRaceWorkflow:
    # Fan a void Activity out across many ids without blocking between sends.
    # asyncio.gather is the equivalent of Promise.allOf — wait for every branch.
    @workflow.run
    async def notify_everyone(self, user_ids: list[str]) -> None:
        opts = dict(start_to_close_timeout=timedelta(minutes=1))
        sends = [workflow.execute_activity(send, uid, **opts) for uid in user_ids]
        await asyncio.gather(*sends)

    # asyncio.FIRST_COMPLETED is the equivalent of Promise.anyOf: continue as
    # soon as the FIRST branch finishes (primary vs fallback provider).
    async def first_to_answer(self, query: str) -> str:
        opts = dict(start_to_close_timeout=timedelta(minutes=1))
        primary = asyncio.ensure_future(workflow.execute_activity(ask_primary, query, **opts))
        fallback = asyncio.ensure_future(workflow.execute_activity(ask_fallback, query, **opts))
        done, _ = await workflow.wait(
            [primary, fallback], return_when=asyncio.FIRST_COMPLETED
        )
        return next(iter(done)).result()

File: examples/02-reliability/java/async_procedure_and_race.java
class AsyncProcedureAndRaceWorkflow {
  private final NotificationActivities notify =
      Workflow.newActivityStub(
          NotificationActivities.class,
          ActivityOptions.newBuilder().setStartToCloseTimeout(Duration.ofMinutes(1)).build());

  // Async.procedure: fire a void Activity without blocking the Workflow loop.
  // (Async.function is for Activities that return a value.)
  void notifyEveryone(List<String> userIds) {
    List<Promise<Void>> sends =
        userIds.stream().map(id -> Async.procedure(notify::send, id)).toList();
    Promise.allOf(sends).get();
  }

  // Promise.anyOf: continue as soon as the FIRST branch finishes
  // (e.g. primary vs fallback provider). allOf waits for every branch.
  String firstToAnswer(String query) {
    Promise<String> primary = Async.function(notify::askPrimary, query);
    Promise<String> fallback = Async.function(notify::askFallback, query);
    Promise.anyOf(primary, fallback).get();
    return primary.isCompleted() ? primary.get() : fallback.get();
  }
}
